cmd/goal/abi/abi.go

```diff
@@ -0,0 +1,1193 @@
+func MakeUint(value *big.Int, size uint16) (Value, error) {
+	typeUint, err := MakeUintType(size)
+	if err != nil {
+		return Value{}, err
+	}
+	upperLimit := big.NewInt(0).Lsh(big.NewInt(1), uint(size))
+	if value.Cmp(upperLimit) >= 0 {
+		return Value{}, fmt.Errorf("passed value larger than uint size %d", size)
+	}
+	return Value{
+		valueType: typeUint,
+		value:     value,
+	}, nil
+}
+
+// MakeUfixed takes a big rational number representation, a type size, and a type precision,
+// and returns an ABI Value of ABI UFixed<size>x<precision>
+func MakeUfixed(value *big.Rat, size uint16, precision uint16) (Value, error) {
+	ufixedValueType, err := MakeUFixedType(size, precision)
+	if err != nil {
+		return Value{}, nil
+	}
+	denomSize := big.NewInt(0).Exp(
+		big.NewInt(10), big.NewInt(int64(precision)),
+		nil,
+	)
+	numUpperLimit := big.NewInt(0).Lsh(big.NewInt(1), uint(size))
+	ufixedLimit := big.NewRat(1, 1).SetFrac(numUpperLimit, denomSize)
+	if value.Denom().Cmp(denomSize) > 0 {
+		return Value{}, fmt.Errorf("value precision overflow")
+	}
+	if value.Cmp(big.NewRat(0, 1)) < 0 || value.Cmp(ufixedLimit) >= 0 {
+		return Value{}, fmt.Errorf("ufixed value out of scope")
+	}
+	return Value{
+		valueType: ufixedValueType,
+		value:     value,
+	}, nil
+}
+
+// MakeString takes a string and returns an ABI String type Value.
+func MakeString(value string) Value {
+	return Value{
+		valueType: MakeStringType(),
+		value:     value,
+	}
+}
+
+// MakeByte takes a byte and returns an ABI Byte type value.
+func MakeByte(value byte) Value {
+	return Value{
+		valueType: MakeByteType(),
+		value:     value,
+	}
+}
+
+// MakeAddress takes an [32]byte array and returns an ABI Address type value.
+func MakeAddress(value [32]byte) Value {
+	return Value{
+		valueType: MakeAddressType(),
+		value:     value,
+	}
+}
+
+// MakeDynamicArray takes an array of ABI value of elemType,
+// and returns an ABI dynamic length array value.
+func MakeDynamicArray(values []Value, elemType Type) (Value, error) {
+	for i := 0; i < len(values); i++ {
+		if !values[i].valueType.Equal(elemType) {
+			return Value{}, fmt.Errorf("type mismatch: %s and %s",
+				values[i].valueType.String(), elemType.String())
+		}
+	}
+	return Value{
+		valueType: MakeDynamicArrayType(elemType),
+		value:     values,
+	}, nil
+}
+
+// MakeStaticArray takes an array of ABI value of elemType,
+// and returns an ABI static length array value.
+func MakeStaticArray(values []Value, elemType Type) (Value, error) {
+	for i := 0; i < len(values); i++ {
+		if !values[i].valueType.Equal(elemType) {
+			return Value{}, fmt.Errorf("type mismatch: %s and %s",
+				values[i].valueType.String(), elemType.String())
+		}
+	}
+	return Value{
+		valueType: MakeStaticArrayType(elemType, uint16(len(values))),
+		value:     values,
+	}, nil
+}
+
+// MakeTuple takes an array of ABI values and an array of ABI types,
+// and returns an ABI tuple value.
+func MakeTuple(values []Value, tupleType []Type) (Value, error) {
```

Comment:
I think so long as we are passing non-empty array/tuple to make function, we can definitely get away with additional type argument. I specified that the input should have length within `[1, 2^16)`, or it will throw error.